### PR TITLE
Updated the BoxVertex format for gui plugins

### LINKED ISSUE #8137: Spine extension looks strange and crashes on Defold 1.6.1 Beta
https://forum.defold.com/t/defold-1-6-1-beta/74460/6

Git bisect pin pointed it it to this [PR](https://github.com/defold/defold/pull/8037), which changed the file format for the gui box vertex format.

### PR DESCRIPTION
**Pull request by JCash**
## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 1]
```diff
@@ -1255,9 +1254,9 @@ namespace dmGameSystem
 
             // Ideally, dmBuffer would support dynamic arrays, but for now this is what we do
             dmArray<uint8_t> node_vertices;
-            type->m_GetVertices(&nodectx, DM_ARRAY_SIZE(boxvertex_stream_decl), boxvertex_stream_decl, struct_size, node_vertices);
+            type->m_GetVertices(&nodectx, gui_world->m_BoxVertexStreamDeclarationCount, gui_world->m_BoxVertexStreamDeclaration, gui_world->m_BoxVertexStructSize, node_vertices);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:1257]: The plugin didn't know about the data format change...

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 2]
```diff
@@ -1255,9 +1254,9 @@ namespace dmGameSystem
...
 
-            uint32_t node_vertex_count = node_vertices.Size() / struct_size;
+            uint32_t node_vertex_count = node_vertices.Size() / gui_world->m_BoxVertexStructSize;
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:1259]: ...nor did we count the size properly here.

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 3]
```diff
@@ -183,6 +183,16 @@ namespace dmGameSystem
         FillAttribute(gui_world->m_ParticleAttributeInfos.m_Infos[2], VERTEX_STREAM_COLOR,      dmGraphics::VertexAttribute::SEMANTIC_TYPE_COLOR,      4);
         FillAttribute(gui_world->m_ParticleAttributeInfos.m_Infos[2], VERTEX_STREAM_PAGE_INDEX, dmGraphics::VertexAttribute::SEMANTIC_TYPE_PAGE_INDEX, 1);
 
+        // Another way would be to use the vertex declaration, but that currently doesn't have an api
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:186]: Moved the code closer to the actual box vertex format, to keep them in sync more easily.

**Comment by AGulev** [comment]
@JCash I think you wanted to assign @Jhonnyg as a reviewer )

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 4]
```diff
@@ -183,6 +183,16 @@ namespace dmGameSystem
...
+        // and the buffer is well suited for this.
+        gui_world->m_BoxVertexStreamDeclarationCount = 4;
+        gui_world->m_BoxVertexStreamDeclaration = new dmBuffer::StreamDeclaration[gui_world->m_BoxVertexStreamDeclarationCount];
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:189]: This is kept around forever, which is probably fine.

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 5]
```diff
@@ -183,6 +183,16 @@ namespace dmGameSystem
...
+        gui_world->m_BoxVertexStreamDeclaration[0] = {VERTEX_STREAM_POSITION,   dmBuffer::VALUE_TYPE_FLOAT32, 3};
+        gui_world->m_BoxVertexStreamDeclaration[1] = {VERTEX_STREAM_TEXCOORD0,  dmBuffer::VALUE_TYPE_FLOAT32, 2};
+        gui_world->m_BoxVertexStreamDeclaration[2] = {VERTEX_STREAM_COLOR,      dmBuffer::VALUE_TYPE_FLOAT32, 4};
+        gui_world->m_BoxVertexStreamDeclaration[3] = {VERTEX_STREAM_PAGE_INDEX, dmBuffer::VALUE_TYPE_FLOAT32, 1};
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:193]: This looks almost exactly as our HVertexStreamDeclaration, except that the struct is not hidden. Maybe we want to merge them at some point?

**Review by Jhonnyg — APPROVED** [review]
👍

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 4 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:189]: Yeah, it is currently passed to the plugins each frame.

File: engine/gamesys/src/gamesys/components/comp_gui.cpp (hunk 5 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/components/comp_gui.cpp:193]: That's definitely the idea in the near future (I hope)

**Pull request merged**